WordPress/gutenberg
# Blocks: Stabilise `role` attribute property

## What?


Stablises the `_experimentalRole` property of block attributes and ~all~ some related selectors and utils.

**Note**: `__experimentalHasContentRoleAttribute` has been made a private selector. Full backwards compatibility will remain until WP 6.8.

## Why?


This has been in use [for 2 years and has not be stablised](https://github.com/WordPress/gutenberg/pull/43037). With all the work on `contentOnly` going on it makes sense to look to stablise the APIs in order that block authors can be confident in making use of them.

## How?


- Adds new stable versions of functions
- Makes `__experimentalHasContentRoleAttribute` a private selector.
- Proxies `__experimental` prefixed versions to new versions and throws `depreciated()` warning
- Ensures stable versions backwards compatible with `__experimentalRole` prop.
- Removes all internal usage of `__experimental` prefixed functions and attribute props.

## Testing Instructions





- Site Editor
- Open dev tools Console
- Type the following to test the `hasContentRoleAttribute` API
```
wp.data.select('core/blocks').__experimentalHasContentRoleAttribute('core/group')
// false

wp.data.select('core/blocks').__experimentalHasContentRoleAttribute('core/heading')
// true
```

- Check deprecated message is as expected.
- Type the following to test the `__experimentalGetBlockAttributesNamesByRole` API:
```
wp.blocks.getBlockAttributesNamesByRole('core/paragraph','content')
// 'content'

wp.blocks.__experimentalGetBlockAttributesNamesByRole('core/paragraph','content')
// 'content'
```

- The call to the experimental method should throw a deprecated notice.

- Also check Editor doesn't crash when moving in / out of Zoom Out mode or [the new "Edit" mode](https://github.com/WordPress/gutenberg/pull/65204).

### Registering a Block

- Post Editor
- Add a title
- Add a paragraph block
- Open dev tools console
- Type the following to register a block using the __experimentalRole: 'local':

```js
wp.blocks.registerBlockType('my-plugin/content-only-test-block', {
    title: 'Content Only Test Block',
    category: 'common',
    attributes: {
        title: {
            type: 'string',
            __experimentalRole: 'local',
        },
    },
    edit: function(props) {
        var title = props.attributes.title;

        return wp.element.createElement(
            'input',
            {
                type: 'text',
                value: title,
                onChange: function(event) {
                    props.setAttributes({ title: event.target.value });
                },
                placeholder: 'Enter title'
            }
        );
    },
    save: function(props) {
        var title = props.attributes.title;
        return wp.element.createElement('h2', null, title);
    },
});
```
- Add the `Content Only Test Block` to the post and **type something into the input**
- Save the post
- See wanring in console about using the deprecated prop.

### Testing Instructions for Keyboard


## Screenshots or screencast
